emoji_presentation_scanner.rl

```diff
@@ -72,6 +72,8 @@ emoji_keycap_sequence = KEYCAP_BASE VS16 COMBINING_ENCLOSING_KEYCAP;
 
 emoji_zwj_element =  emoji_presentation_sequence | emoji_modifier_sequence | any_emoji;
 
+emoji_zwj_sequence_with_vs = (emoji_zwj_element ZWJ)* emoji_presentation_sequence (ZWJ emoji_zwj_element)*;
```

Comment:
Can this rule be further reduced and more analogous to the shape of `emoji_zwj_sequence`? The shape of this rule is a bit different from `emoji_zwj_sequence`: There's an extra repetition of (emoji_zwj_element ZWJ)* at the beginning. And the quantifier changes from `+` to `*`.

Reply:
Not sure, we need at least one `emoji_zwj_element` in the rule:

`emoji_zwj_sequence = emoji_zwj_element ( ZWJ emoji_zwj_element )+;` 

to have `VS16`  in order to set `has_vs=true`.

Reply:
I am afraid I don't understand what you mean. Can you explain in more detail what you meant? The regular production for `emoji_zwj_sequence` has "at least one emoji_zwj_element in the rule" already. Do you mean, we need at least one `emoji_presentation_sequence` in the new `emoji_zwj_sequence_with_vs` rule? 

Could you try to minimize the rule further and update PR?

Reply:
Yes, the regular production for `emoji_zwj_sequence = emoji_zwj_element ( ZWJ emoji_zwj_element )+;` has "at least one `emoji_zwj_element` in the rule" already, but I'm saying at least one of  `emoji_zwj_element` should include `VS16`.  So yes, we need at least one `emoji_presentation_sequence` in the new `emoji_zwj_sequence_with_vs` rule.

Reply:
Could not come up with anything smaller, do you have any ideas?

Reply:
So, the bug talks about where with the VS16 and `font-variant-emoji: text`, the rendering fails and is shown as tofu, even though the sequence contains a VS16 that should not be touched/modified by the property, right?
`U+2764, U+FE0F, U+200D, U+1F525`

So you're looking for a rule that recognizes, that at least the first part of the ZWJ sequence, here the `U+2764, U+FE0F` is a VS ZWJ sequence, right?

I think you might need to split 
`emoji_zwj_element` into two productions, one for `emoji_zwj_vs_element`, and one for `emoji_zwj_element` without VS. 

And then also split `emoji_zwj_sequence` into two productions of the same form, one for VS-containing ZWJ sequences based on `emoji_zwj_vs_element`, and one without based on non-VS `emoji_zwj_element`, and then add the one without to the `emoji_presentation` production, and the one with VS to `emoji_run_with_vs`.

(untested)

Reply:
Now, thinking about it again, I find spec a bit unclear in cases when zwj does not have VS. We probably don't want to apply font-variant-emoji on them as well. But according to the [CSS Fonts Module 4 spec](https://www.w3.org/TR/css-fonts-4/#font-variant-emoji-prop:~:text=Only%20the%20code%20points%20listed%20by%20Unicode%20as%20contributing%20to%20a%20Unicode%20emoji%20presentation%20sequence%20are%20affected%20by%20this%20property), in following zwj sequence: 

`1F408 200D 2B1B; RGI_Emoji_ZWJ_Sequence; black cat   # E13.0  [1] (🐈‍⬛)`

How will `font-variant-emoji` work? Should it work as `1F408 FE0E 200D 2B1B` or should we just ignore `1F408 200D 2B1B`? The way I interpret the spec now is that it should work as `1F408 FE0E 200D 2B1B`, but I think it might make sense to not apply font-variant-emoji on all ZWJ sequences. I will post it in the issue comment.

Reply:
> I will post it in the issue comment.

In the Chromium bug or as a comment in a CSS WG issue?

Yes, I agree with you that we to agree what should happen with those. Unicode RGI (recommended for general interchange) ZWJ sequences are here:
https://www.unicode.org/Public/emoji/latest/emoji-zwj-sequences.txt

This list contains sequences with and without VS16. The logic seems to be that if the participating emoji is EmojiPresentation=No, the VS16 is included (U+2764, U+27A1). 

2764 FE0F 200D 1F525      "heart on fire" is on the list, so here font-variant-emoji should not override the variation selector. But there are sequences that do not contains a VS, and downgrading participating emoji in the sequence to VS15 would be somewhat odd, as that would make it a non-RGI sequence and likely split the sequence in presentation.

Reply:
Yes, exactly, thanks.

>In the Chromium bug or as a comment in a CSS WG issue?

I meant on Chromium, but will open a new CSS WG issue for that as well, will cc you.

Reply:
> 2764 FE0F 200D 1F525

In principle U+1F525 (FIRE) is in this sequence without a variation selector so font-variant-emoji would apply here and change the sequence.

Reply:
So let's say we have:

```<div style="font-variant-emoji: text">&#x1F408;&#x200D;&#x2B1B;`</div>```

Then the way I interpret spec it should be the same as if:

```<div>&#x1F408;&#xFE0E;&#x200D;&#x2B1B;`</div>```

since U+1F408 U+FE0E is a valid variation sequence, compare https://www.unicode.org/emoji/charts/emoji-variants.html

Reply:
Sorry, I'm a bit confused, are we at the same page on filing CSS WG issue then?